src/main/scala/org/graphframes/lib/ConnectedComponents.scala

```diff
@@ -346,9 +340,20 @@ private object ConnectedComponents extends Logging {
       ee.persist(StorageLevel.MEMORY_AND_DISK)
 
       // test convergence
-      val currSum = ee.select(sum(col(SRC))).as[Long].first()
+
+      // Taking the sum in DecimalType to preserve precision.
+      // We use 20 digits for long values and Spark SQL will add 10 digits for the sum.
+      // It should be able to handle 100 billion edges without overflow.
+      val (currSum, cnt) = ee.select(sum(col(SRC).cast(DecimalType(20, 0))), count("*")).rdd
+        .map { r =>
+          (r.getAs[BigDecimal](0), r.getLong(1))
+        }.first()
+      if (cnt != 0L && currSum == null) {
+        throw new ArithmeticException("The total sum of current assignments exceeded limit 1e30.")
```

Comment:
1e30 is not the number of edges. It is the sum of component IDs assigned. The component IDs could be as big as 1e19 (Long.MaxValue). So we can handle ~200 billion edges. I will update the message to make the message clearer.